### DISCUSSION TITLE
Clarification about iterating with MidiFile object

### DISCUSSION THREAD

**Discussion by bixb922** [mido/mido#554] [General]
I want to iterate over the objects in newly created a MidiFile object. Based on the sample code at https://mido.readthedocs.io/en/stable/files/midi.html I wrote the following script:
```py
from mido import Message, MidiFile, MidiTrack
# Create MIDI file in memory according to example at 
# https://mido.readthedocs.io/en/stable/files/midi.html
mid = MidiFile()
track = MidiTrack()
mid.tracks.append(track)

track.append(Message('program_change', program=12, time=0))
track.append(Message('note_on', note=64, velocity=64, time=32))
track.append(Message('note_off', note=64, velocity=127, time=32))

# Iterate over messages in mid object following  example at same page
for m in mid:
  print("In-memory file, message", m )

# Work around: save MIDI file to disk and reload:
print("Saving MIDI file and reloading from disk")
mid.save("testsong2.mid")
mid2 = MidiFile( "testsong2.mid")

for m in mid2:
  print("Saved MIDI file, message", m )
```

The output I get is:
```
py testmido.py
In-memory file, message MetaMessage('end_of_track', time=0)
Saving MIDI file and reloading from disk
Saved MIDI file, message program_change channel=0 program=12 time=0
Saved MIDI file, message note_on channel=0 note=64 velocity=64 time=0.03333333333333333
Saved MIDI file, message note_off channel=0 note=64 velocity=127 time=0.03333333333333333
Saved MIDI file, message MetaMessage('end_of_track', time=0)
```
What I expected is that the unsaved MIDI file also allows iterating over the messages, but that first loop only shows the end-of-track message. Iterating over tracks and then over messages by track yield the two messages. According to a program I developed and tested about a year ago, it seems to me that iterating over a newly created MidiFile object was supported. As a workaround, I can certainly save and reload the MIDI file.

This is with MIDO 1.3.0 and Python 3.11.4 on Windows 11.

Please could anyone point out what I am doing wrong?

**Comment by rdoursenaud** [comment]
Hi and thanks for the detailed report!

It's a regression that's been introduced by c5b98f4594d74bb15b0d92c8b0e3546d30e421b2 / #470  which optimizes file loading from disk.

As a workaround, add the following before iterating over messages.
```py
from mido.midifiles.tracks import merge_tracks

mid.merged_track = merge_tracks(mid.tracks)
```

I'm redesigning Standard MIDI Files handling for version 2 which will not have these side effects.

**Comment by bixb922** [comment]
Thank you very much for the prompt response and the workaround!

**Comment by akx** [comment]
To follow up: Doing

```
track.append(Message('program_change', program=12, time=0))
track.append(Message('note_on', note=64, velocity=64, time=32))
track.append(Message('note_off', note=64, velocity=127, time=32))
mid.tracks.append(track)
```
instead will fix this (i.e. add the track only after you've mutated it, so the internal `merged_tracks` cache gets cleared).

With #565 having landed, this particular behavior is fixed (not released yet), but if you happen to mutate a track after having iterated over the file once, the new events will not appear when you iterate over it again.
Similarly, after #565 is released, you can clear the cache with `del mid.merged_tracks`, and it will be recomputed the next time you iterate over the file.

**Reply by bixb922** [reply]
Thank you!!
